### PR TITLE
Further progress on right orthogonal calculus

### PR DESCRIPTION
**Pull request by TashiWalde**
1) Right orthogonal maps are closed under composition and right cancel.

2) Right orthogonal maps are closed under relative products (=pullback).

Point 2 was surprisingly tricky. Maybe there was an easier way to do this. I did this by setting up a theory of relative extension types and generalized relative extension types which might be of independent interest.

The two main steps are:

3) A map is right orthogonal if and only if all its relative extension types are contractible if and only if all its generalized relative extension types are contractible.

4) Each relative extension type of the pulled back map appears as retract of a generalized relative extension type of the original map.

Notes:
- Depends on #111, where relative products are introduced.
- Partially addresses #81.
- Added a few minor utilities.

### CODE REVIEW

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 1]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
 
 #end homotopy-cartesian-horizontal-calculus
 ```
+
+## Fiber products
+
+Given two type families `B C : A → U`, we can form their fiberwise product.
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:417]:
```suggestion
Given two type families `B C : A → U`, we can form their **fiberwise product**.
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 2]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+  : A → U
+  :=
+    \ a → product (B a) (C a)
+
+#def first-fiberwise-product
+  ( A : U)
+  ( B C : A → U)
+  ( a : A)
+  : fiberwise-product A B C a → B a
+  := \ (b,_) → b
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:432]:
```suggestion
  := \ (b , _) → b
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 3]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def second-fiberwise-product
+  ( A : U)
+  ( B C : A → U)
+  ( a : A)
+  : fiberwise-product A B C a → C a
+  := \ (_,c) → c
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:439]:
```suggestion
  := \ (_ , c) → c
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 4]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#variable A : U
+#variable B : U
+#variable β : B → A
+#variable C : U
+#variable γ : C → A
+
+#def relative-product
+  : U
+  := Σ ( (b, c) : product B C) , β b = γ c
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:455]:
```suggestion
  := Σ ( (b , c) : product B C) , (β b = γ c)
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 5]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def second-fiberwise-product
+  ( A : U)
+  ( B C : A → U)
+  ( a : A)
+  : fiberwise-product A B C a → C a
+  := \ (_,c) → c
+```
+
+Given two maps `B → A` and `C → A`, we can form the relative product over `A`.
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:442]:
```suggestion
Given two maps `B → A` and `C → A`, we can form the **relative product** over `A`.
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 6]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+#variable C : U
+#variable γ : C → A
+
+#def relative-product
+  : U
+  := Σ ( (b, c) : product B C) , β b = γ c
+
+#def first-relative-product uses (A B β C γ)
+  : relative-product → B
+  := \ ((b , _), _) → b
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:459]:
```suggestion
  := \ ((b , _) , _) → b
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 7]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def second-relative-product uses (A B β C γ)
+  : relative-product → C
+  := \ ((_ , c), _) → c
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:463]:
```suggestion
  := \ ((_ , c) , _) → c
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 8]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def homotopy-relative-product uses (A B C)
+  ( (bc, p) : relative-product )
+  : β (first-relative-product (bc,p)) = γ (second-relative-product (bc,p))
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:467]:
```suggestion
  ( (bc , p) : relative-product )
  : β (first-relative-product (bc , p)) = γ (second-relative-product (bc , p))
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 9]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+product of all fibers.
+
+```rzk
+#def fiber-product
+  : U
+  := total-type A (fiberwise-product A (fib B A β) (fib C A γ))
+
+#def fiber-product-unpacked
+  : fiber-product
+  = ( Σ (a : A), (product (fib B A β a) (fib C A γ a)))
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:481]:
```suggestion
  = ( Σ (a : A) , (product (fib B A β a) (fib C A γ a)))
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 10]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+  := refl
+
+#def first-fiber-product uses (A B β C γ)
+  : fiber-product → B
+  := \ (_, ((b, _), _ )) → b
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:486]:
```suggestion
  := \ (_ , ((b , _) , _ )) → b
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 11]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def second-fiber-product uses (A B β C γ)
+  : fiber-product → C
+  := \ (_, (_, (c, _))) → c
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:490]:
```suggestion
  := \ (_ , (_ , (c , _))) → c
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 12]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+
+#def homotopy-fiber-product uses (A B C)
+  : ( abpcq : fiber-product )
+  → β (first-fiber-product abpcq) = γ (second-fiber-product abpcq)
+  :=
+    \ ( a, ((b, p), (c,q))) →
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:496]:
```suggestion
    \ ( a , ((b , p) , (c , q))) →
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 13]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+      zig-zag-concat A (β b) a (γ c) p q
+
+#def relative-fiber-product uses (B C)
+  ( (a, ((b, p), (c,q))) : fiber-product )
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:500]:
```suggestion
  ( (a , ((b , p) , (c , q))) : fiber-product)
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 14]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+  : relative-product
+  := ( ( b , c) , zig-zag-concat A (β b) a (γ c) p q)
+
+#def fiber-relative-product uses ( A B β C)
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:504]:
```suggestion
#def fiber-relative-product uses (A B β C)
```

File: src/hott/11-homotopy-pullbacks.rzk.md [hunk 15]
```diff
@@ -411,3 +411,134 @@ from composition and cancelling laws for equivalences.
...
+#def is-id-relative-fiber-relative-product
+  ( bce : relative-product)
+  : relative-fiber-product (fiber-relative-product bce) = bce
+  := refl
+
+#def is-id-fiber-relative-fiber-product
+  : ( abpcq : fiber-product)
+  → ( fiber-relative-product (relative-fiber-product abpcq)) = abpcq
+  :=
+  \ (a', (bq', cq')) →
```
Comment (fredrik-bakke) [inline src/hott/11-homotopy-pullbacks.rzk.md:518]:
```suggestion
  \ (a' , (bq' , cq')) →
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 16]
```diff
@@ -48,7 +48,7 @@ restriction map `(ψ → A) → (ϕ → A)`, which we can view as the types of
   :=
     \ τ → ( τ, refl)
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:49]:
```suggestion
    \ τ → ( τ , refl)
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 17]
```diff
@@ -63,17 +63,24 @@ restriction map `(ψ → A) → (ϕ → A)`, which we can view as the types of
...
-    \ th → first (extension-type-weakening-section σ th)
+    \ th → first (section-extension-type-weakening' σ th)
+
+#def has-section-extension-type-weakening
+  ( σ : (t : ϕ) → A t)
+  : has-section (extension-type σ) (homotopy-extension-type σ)
+      (extension-type-weakening-map σ)
+  :=
+    ( extension-strictification σ,
+      \ th → ( second (section-extension-type-weakening' σ th)))
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:74]:
```suggestion
    ( extension-strictification σ
    , \ th → ( second (section-extension-type-weakening' σ th)))
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 18]
```diff
@@ -63,17 +63,24 @@ restriction map `(ψ → A) → (ϕ → A)`, which we can view as the types of
...
 #def is-equiv-extension-type-weakening
   ( σ : (t : ϕ) → A t)
   : is-equiv (extension-type σ) (homotopy-extension-type σ)
       (extension-type-weakening-map σ)
   :=
-    ( ( extension-strictification σ ,
-        \ _ → refl),
-      ( extension-strictification σ,
-        \ th → ( second (extension-type-weakening-section σ th))))
+    ( ( extension-strictification σ, \ _ → refl)
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:82]:
```suggestion
    ( ( extension-strictification σ , \ _ → refl)
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 19]
```diff
@@ -116,7 +123,6 @@ This equivalence is functorial in the following sense:
       , is-equiv-extension-type-weakening I ψ ϕ A (\ t → α t (σ' t))
       )
     )
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:125]:
```suggestion
      , is-equiv-extension-type-weakening I ψ ϕ A (\ t → α t (σ' t))))
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 20]
```diff
@@ -1057,3 +1063,191 @@ $\left \langle_{t : I |\psi} f(t) = a'(t) \biggr|^\phi_{\lambda t.refl} \right\r
...
+    ( homotopy-extension-type I ψ ϕ A σ)
+    ( pointwise-homotopy-extension-type σ)
+  :=
+    total-equiv-family-equiv
+    ( (t : ψ) → A t)
+    ( \ τ → (\ t → τ t) =_{ (t : ϕ) → A t} σ)
+    ( \ τ → (t : ϕ) → (τ t = σ t))
+    ( \ τ →
+      equiv-ExtExt extext I (\ t → ϕ t) (\ _ → BOT) (\ t → A t)
+      ( \ _ → recBOT) ( \ t → τ t) ( σ))
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:1098]:
```suggestion
      equiv-ExtExt extext I (\ t → ϕ t) (\ _ → BOT) (\ t → A t)
      ( \ _ → recBOT) (\ t → τ t) σ)
```

File: src/simplicial-hott/04-right-orthogonal.rzk.md [hunk 21]
```diff
@@ -52,6 +53,81 @@ orthogonal_ to the map `α`, if `α : A' → A` is right orthogonal to `ϕ ⊂ 
...
+        ( (t : ψ) → A' [ϕ t ↦ σ' t])
+        ( (t : ψ) → A [ϕ t ↦ α (σ' t)])
+        ( \ τ' t → α (τ' t))
+        ( τ))
+      ( relative-extension-type I ψ ϕ
+        ( \ _ → A') (\ _ → A) (\ _ → α) σ' τ)
+      ( equiv-relative-extension-type-fib extext I ψ ϕ
+        ( \ _ → A') (\ _ → A) (\ _ → α) σ' τ)
+      ( are-contr-relext-α σ' τ)
+    )
```
Comment (fredrik-bakke) [inline src/simplicial-hott/04-right-orthogonal.rzk.md:94]:
```suggestion
      ( are-contr-relext-α σ' τ))
```

File: src/simplicial-hott/04-right-orthogonal.rzk.md [hunk 22]
```diff
@@ -302,6 +368,218 @@ For any two shapes `ϕ, ψ ⊂ I`, if `ϕ ∩ ψ ⊂ ϕ` is left orthogonal to
...
+  :=
+    ( \ t → ( (τA' t, τB t) , hA t)
+    , \ t → refl)
+
+#def general-relative-extension-type-relative-extension-type-pullback
+  ( σB' : ϕ → relative-product A A' α B f)
+  ( τB : (t : ψ) → B [ϕ t ↦ second-relative-product A A' α B f (σB' t)])
+  ( (τB', hB)
+    : relative-extension-type I ψ ϕ
+      ( \ _ → relative-product A A' α B f) ( \ _ → B)
```
Comment (fredrik-bakke) [inline src/simplicial-hott/04-right-orthogonal.rzk.md:464]:
```suggestion
      ( \ _ → relative-product A A' α B f) (\ _ → B)
```

File: src/simplicial-hott/04-right-orthogonal.rzk.md [hunk 23]
```diff
@@ -302,6 +368,218 @@ For any two shapes `ϕ, ψ ⊂ I`, if `ϕ ∩ ψ ⊂ ϕ` is left orthogonal to
...
+    ( \ _ → second-relative-product A A' α B f)
+    ( σB') ( τB)
+  :=
+    ( \ t → ( (τA' t, τB t) , hA t)
+    , \ t → refl)
+
+#def general-relative-extension-type-relative-extension-type-pullback
+  ( σB' : ϕ → relative-product A A' α B f)
+  ( τB : (t : ψ) → B [ϕ t ↦ second-relative-product A A' α B f (σB' t)])
+  ( (τB', hB)
```
Comment (fredrik-bakke) [inline src/simplicial-hott/04-right-orthogonal.rzk.md:462]:
```suggestion
  ( (τB' , hB)
```

**Review by fredrik-bakke — COMMENTED** [review]
Okay, I've had enough of making all these formatting corrections, and I'm sure you are done with formatting code and seeing my comments as well. I'll write an autoformatting script over the weekend so we can move on to more productive things.

File: src/simplicial-hott/04-right-orthogonal.rzk.md [hunk 24]
```diff
@@ -52,6 +53,81 @@ orthogonal_ to the map `α`, if `α : A' → A` is right orthogonal to `ϕ ⊂ 
...
+
+#def has-contr-general-relative-extension-types-is-right-orthogonal-to-shape
+  uses (extext)
+  ( is-orth-α : is-right-orthogonal-to-shape I ψ ϕ A' A α)
+  : has-contr-general-relative-extension-types I ψ ϕ
+      ( \ _ → A') (\ _ → A) (\ _ → α)
+  :=
+    has-contr-relative-extension-types-generalize extext I ψ ϕ
+      ( \ _ → A') (\ _ → A) (\ _ → α)
+    ( has-contr-relative-extension-types-is-right-orthogonal-to-shape is-orth-α)
```
Comment (fredrik-bakke) [inline src/simplicial-hott/04-right-orthogonal.rzk.md:126]:
```suggestion
      ( has-contr-relative-extension-types-is-right-orthogonal-to-shape is-orth-α)
```

File: src/simplicial-hott/03-extension-types.rzk.md [hunk 25]
```diff
@@ -1057,3 +1063,191 @@ $\left \langle_{t : I |\psi} f(t) = a'(t) \biggr|^\phi_{\lambda t.refl} \right\r
...
+If all ordinary relative extension types are contractible, then also all
+generalized ones.
+
+```rzk
+#def has-contr-relative-extension-types
+  : U
+  :=
+    ( σ' : (t : ϕ) → A' t)
+  → ( τ : (t : ψ) → A t [ϕ t ↦ α t (σ' t)])
+  → ( is-contr (relative-extension-type I ψ ϕ A' A α σ' τ))
```
Comment (fredrik-bakke) [inline src/simplicial-hott/03-extension-types.rzk.md:1194]:
```suggestion
  ( ( σ' : (t : ϕ) → A' t)
  → ( τ : (t : ψ) → A t [ϕ t ↦ α t (σ' t)])
  → ( is-contr (relative-extension-type I ψ ϕ A' A α σ' τ)))
```

**Review by fredrik-bakke — APPROVED** [review]
Let me say I think this PR looks excellent overall.

**Comment by TashiWalde** [comment]
@fredrik-bakke I have dismissed (marked as resolved) some of your suggestions which should be addressed in the PR #111 on which this one builds.

**Comment by fredrik-bakke** [comment]
In case you missed it, there's a single comment left to resolve. After that, this PR seems ready to merge to me.

**Review by fredrik-bakke — APPROVED** [review]

**Comment by fredrik-bakke** [comment]
> Okay, I've had enough of making all these formatting corrections, and I'm sure you are done with formatting code and seeing my comments as well. I'll write an autoformatting script over the weekend so we can move on to more productive things.

I had a chat with Nikolai before the weekend. He and his student will try to set up basic autoformatting functionality built into the Rzk ecosystem instead, so the autoformatting will have to wait a little.

**Pull request merged**